framework/src/org/checkerframework/framework/type/AsSuperVisitor.java

```diff
@@ -0,0 +1,807 @@
+import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.InternalUtils;
+import org.checkerframework.javacutil.TypesUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Types;
+
+/**
+ * Implements asSuper {@link AnnotatedTypes#asSuper(AnnotatedTypeFactory, AnnotatedTypeMirror,
+ * AnnotatedTypeMirror)}
+ */
+public class AsSuperVisitor extends AbstractAtmComboVisitor<AnnotatedTypeMirror, Void> {
+
+    private final Types types;
+    private final AnnotatedTypeFactory annotatedTypeFactory;
+
+    public AsSuperVisitor(AnnotatedTypeFactory annotatedTypeFactory) {
+        this.annotatedTypeFactory = annotatedTypeFactory;
+        types = annotatedTypeFactory.types;
+    }
+
+    /**
+     * Implements asSuper {@link AnnotatedTypes#asSuper(AnnotatedTypeFactory, AnnotatedTypeMirror,
+     * AnnotatedTypeMirror)}
+     *
+     * @param type      Type from which to copy annotations
+     * @param superType a type whose erased Java type is a supertype of {@code type}'s erased Java
+     *                  type.
+     * @return {@code superType} with annotations copied from {@code type} and type variables
+     * substituted from {@code type}.
+     */
+    @SuppressWarnings("unchecked")
+    public <T extends AnnotatedTypeMirror> T asSuper(AnnotatedTypeMirror type, T superType) {
+        if (type == null || superType == null) {
+            ErrorReporter.errorAbort("AsSuperVisitor type and supertype cannot be null.");
+            return null; // dead code
+        } else if (type == superType) {
+            return (T) type.deepCopy();
+        }
+        AnnotatedTypeMirror copySuperType = superType.deepCopy(false);
+        AnnotatedTypeMirror result = visit(type.deepCopy(), copySuperType, null);
+
+        if (result == null) {
+            ErrorReporter.errorAbort(
+                    "AsSuperVisitor returned null.\ntype: %s\nsuperType: %s", type, copySuperType);
+        }
+
+        return (T) result;
+    }
+
+    @Override
+    public AnnotatedTypeMirror visit(AnnotatedTypeMirror type, AnnotatedTypeMirror superType,
+                                     Void p) {
+        ensurePrimaryIsCorrectForUnionsAndIntersections(type);
+        return super.visit(type, superType, p);
+    }
+
+    /**
+     * The code in this class is assuming that the primary annotation of an {@link AnnotatedIntersectionType}
+     * is the greatest lower bound of the annotations on its direct super types and that the primary
+     * annotation of an {@link AnnotatedUnionType} is the least upper bound of its alternatives.
+     * This method makes this assumption true.
+     */
+    private void ensurePrimaryIsCorrectForUnionsAndIntersections(AnnotatedTypeMirror type) {
```

Comment:
In general, we need to actually decide what the primary annotation on a union or intersection should mean and then implement that.  So, I choose something sensible so that I could finish asSuper.
>What is the use of primary annotations on Intersection and Union types?

They are only useful so that getAnnotations returns a non-null value.

>If I remember correctly, Java 8 provides no syntax to provide an annotation for the overall type.
Would it prevent accidental additions by preventing primary annotations on Intersection/Union in the AnnotatedTypeMirror subclasses?
How can this property ever be violated if there is no syntax to specify something else?

A checker can set the primary annotation to what ever it likes.